src/PoolManager.sol

```diff
@@ -203,34 +214,62 @@ contract PoolManager is IPoolManager, ProtocolFees, NoDelegateCall, ERC6909Claim
         PoolId id = key.toId();
         _checkPoolInitialized(id);
 
-        key.hooks.beforeSwap(key, params, hookData);
-
-        uint256 feeForProtocol;
-        uint24 swapFee;
-        Pool.SwapState memory state;
-        (delta, feeForProtocol, swapFee, state) = pools[id].swap(
-            Pool.SwapParams({
-                tickSpacing: key.tickSpacing,
-                zeroForOne: params.zeroForOne,
-                amountSpecified: params.amountSpecified,
-                sqrtPriceLimitX96: params.sqrtPriceLimitX96
-            })
-        );
-
-        _accountPoolBalanceDelta(key, delta);
-
-        // the fee is on the input currency
-        unchecked {
-            if (feeForProtocol > 0) {
-                protocolFeesAccrued[params.zeroForOne ? key.currency0 : key.currency1] += feeForProtocol;
+        int128 hookDeltaInSpecified = key.hooks.beforeSwap(key, params, hookData);
+
+        bool exactInput = params.amountSpecified < 0;
+        bool zeroIsSpecified = exactInput == params.zeroForOne;
+
+        // Update the swap amount according to the hook's return, and check that the swap type doesnt change (exact input/output)
+        params.amountSpecified += hookDeltaInSpecified;
+        if (exactInput ? params.amountSpecified > 0 : params.amountSpecified < 0) revert SwapTypeChanged();
+
+        // Account the hook's delta to the hook's address
+        _accountDeltaFor(zeroIsSpecified ? key.currency0 : key.currency1, hookDeltaInSpecified, address(key.hooks));
+
+        if (params.amountSpecified != 0) {
+            uint256 feeForProtocol;
+            uint24 swapFee;
+            Pool.SwapState memory state;
+            (delta, feeForProtocol, swapFee, state) = pools[id].swap(
```

Comment:
It seems so strange to have the actual call to swap (the meat of this function) be nested. I wonder if we could have something like 

if ( specified == 0) { _handleAfterSwap())
swap()
_handleAfterSwap()

or could we handle the case where it equals 0 inside swap like before?

Reply:
i put it inside the pool contract